streams/src/main/java/org/apache/kafka/streams/state/RocksDBWindowStore.java

```diff
@@ -0,0 +1,281 @@
+import java.util.Set;
+import java.util.SimpleTimeZone;
+
+public class RocksDBWindowStore<K, V> implements WindowStore<K, V> {
+
+    public static final long MIN_SEGMENT_INTERVAL = 60 * 1000; // one minute
+
+    private static class Segment extends RocksDBStore<byte[], byte[]> {
+        public final long id;
+
+        Segment(String name, long id) {
+            super(name, WindowStoreUtil.INNER_SERDES);
+            this.id = id;
+        }
+
+        public void destroy() {
+            Utils.delete(dbDir);
+        }
+    }
+
+    private static class RocksDBWindowStoreIterator<V> implements WindowStoreIterator<V> {
+        private final Serdes<?, V> serdes;
+        private final KeyValueIterator<byte[], byte[]>[] iterators;
+        private int index = 0;
+
+        RocksDBWindowStoreIterator(Serdes<?, V> serdes) {
+            this(serdes, WindowStoreUtil.NO_ITERATORS);
+        }
+
+        RocksDBWindowStoreIterator(Serdes<?, V> serdes, KeyValueIterator<byte[], byte[]>[] iterators) {
+            this.serdes = serdes;
+            this.iterators = iterators;
+        }
+
+        @Override
+        public boolean hasNext() {
+            while (index < iterators.length) {
+                if (iterators[index].hasNext())
+                    return true;
+
+                index++;
+            }
+            return false;
+        }
+
+        @Override
+        public V next() {
+            if (index >= iterators.length)
+                throw new NoSuchElementException();
+
+            return serdes.valueFrom(iterators[index].next().value());
+        }
+
+        @Override
+        public void remove() {
+            if (index < iterators.length)
+                iterators[index].remove();
+        }
+
+        public void close() {
+            for (KeyValueIterator<byte[], byte[]> iterator : iterators) {
+                iterator.close();
+            }
+        }
+    }
+
+    private final String name;
+    private final long windowBefore;
+    private final long windowAfter;
+    private final long segmentInterval;
+    private final Segment[] segments;
+    private final Serdes<K, V> serdes;
+    private final SimpleDateFormat formatter;
+
+    private ProcessorContext context;
+    private long currentSegmentId = -1L;
+    private int seqnum = 0;
+
+    public RocksDBWindowStore(String name, long windowBefore, long windowAfter, long retentionPeriod, int numSegments, Serdes<K, V> serdes) {
+        this.name = name;
+        this.windowBefore = windowBefore;
+        this.windowAfter = windowAfter;
+
+        // The retention period must be at least four times as long as the windowSize
+        if ((this.windowBefore + this.windowAfter + 1) * 2 > retentionPeriod)
+            retentionPeriod = (this.windowBefore + this.windowAfter + 1) * 2;
+
+        // The segment interval must be greater than MIN_SEGMENT_INTERVAL
+        this.segmentInterval = Math.max(retentionPeriod / (numSegments - 1), MIN_SEGMENT_INTERVAL);
```

Comment:
Since one of the segments is always empty when it is created, we have to cover the retention period with (numSegment - 1) segments in order to guarantee the retention period.